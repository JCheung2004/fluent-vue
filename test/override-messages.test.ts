import { createLocalVue, mount } from '@vue/test-utils'

import { FluentBundle, FluentResource } from '@fluent/bundle'
import ftl from '@fluent/dedent'

import FluentVue from '../src'

describe('child component', () => {
  let options: any
  let bundleEn: FluentBundle
  let bundleUk: FluentBundle

  beforeEach(() => {
    const localVue = createLocalVue()
    localVue.use(FluentVue)

    bundleEn = new FluentBundle('en-US')
    bundleUk = new FluentBundle('uk-UA')

    const fluent = new FluentVue({
      bundles: [bundleUk, bundleEn]
    })

    options = {
      fluent,
      localVue
    }
  })

  it('can override parent messages', () => {
    // Arrange
    bundleEn.addResource(
      new FluentResource(ftl`
      link = link text
      `)
    )

    bundleUk.addResource(
      new FluentResource(ftl`
      link = текст посилання
      `)
    )

    const child = {
      fluent: {
        'en-US': new FluentResource(ftl`
        link = child link text
        `),
        'uk-UA': new FluentResource(ftl`
        link = текст посилання з під-компонента
        `)
      } as any,
      template: `<div>{{ $t('link') }}</div>`
    }

    const component = {
      components: {
        child
      },
      template: `
      <div>
        {{ $t('link') }}
        <child></child>
      </div>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()

    // Change language
    options.fluent.bundles = [bundleEn, bundleUk]

    expect(mounted).toMatchSnapshot()
  })

  it('updates child language when parent language changes', () => {
    // Arrange
    bundleEn.addResource(
      new FluentResource(ftl`
      link = link text
      `)
    )

    bundleUk.addResource(
      new FluentResource(ftl`
      link = текст посилання
      `)
    )

    const child = {
      fluent: {
        'en-US': new FluentResource(ftl`
        link = child link text
        `),
        'uk-UA': new FluentResource(ftl`
        link = текст посилання з під-компонента
        `)
      } as any,
      template: `<div>{{ $t('link') }}</div>`
    }

    const component = {
      components: {
        child
      },
      template: `
      <div>
        {{ $t('link') }}
        <child></child>
      </div>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()

    // Change language
    options.fluent.bundles = [bundleEn, bundleUk]

    expect(mounted).toMatchSnapshot()
  })

  it('updates grandchild language when parent language changes', () => {
    // Arrange
    bundleEn.addResource(
      new FluentResource(ftl`
      link = link text
      `)
    )

    bundleUk.addResource(
      new FluentResource(ftl`
      link = текст посилання
      `)
    )

    const grandChild = {
      fluent: {
        'en-US': new FluentResource(ftl`
        link = grandchild link text
        `),
        'uk-UA': new FluentResource(ftl`
        link = текст посилання з під-під-компонента
        `)
      } as any,
      template: `<div>{{ $t('link') }}</div>`
    }

    const child = {
      components: {
        grandChild
      },
      fluent: {
        'en-US': new FluentResource(ftl`
        link = child link text
        `),
        'uk-UA': new FluentResource(ftl`
        link = текст посилання з під-компонента
        `)
      } as any,
      template: `
      <div>
        <div>{{ $t('link') }}</div>
        <grand-child></grand-child>
      </div>`
    }

    const component = {
      components: {
        child
      },
      template: `
      <div>
        {{ $t('link') }}
        <child></child>
      </div>`
    }

    // Act
    const mounted = mount(component, options)

    // Assert
    expect(mounted).toMatchSnapshot()

    // Change language
    options.fluent.bundles = [bundleEn, bundleUk]
    expect(mounted).toMatchSnapshot()
  })
})
