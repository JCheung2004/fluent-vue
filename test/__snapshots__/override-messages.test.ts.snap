// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`child component can override parent messages 1`] = `
<div>
  текст посилання
  <div>текст посилання з під-компонента</div>
</div>
`;

exports[`child component can override parent messages 2`] = `
<div>
  link text
  <div>child link text</div>
</div>
`;

exports[`child component updates child language when parent language changes 1`] = `
<div>
  текст посилання
  <div>текст посилання з під-компонента</div>
</div>
`;

exports[`child component updates child language when parent language changes 2`] = `
<div>
  link text
  <div>child link text</div>
</div>
`;

exports[`child component updates grandchild language when parent language changes 1`] = `
<div>
  текст посилання
  <div>
    <div>текст посилання з під-компонента</div>
    <div>текст посилання з під-під-компонента</div>
  </div>
</div>
`;

exports[`child component updates grandchild language when parent language changes 2`] = `
<div>
  link text
  <div>
    <div>child link text</div>
    <div>grandchild link text</div>
  </div>
</div>
`;
